crates/router/src/db/role.rs

```diff
@@ -357,40 +350,71 @@ impl RoleInterface for MockDb {
-    ) -> CustomResult<Vec<storage::Role>, errors::StorageError> {
+        role_name: &str,
+        entity_type: storage::ListRolesByEntityPayload,
+    ) -> CustomResult<storage::Role, errors::StorageError> {
         let roles = self.roles.lock().await;
 
-        let roles_list: Vec<_> = roles
+        let roles_list = roles
             .iter()
-            .filter(|role| {
-                role.merchant_id == *merchant_id
-                    || (role.org_id == *org_id
-                        && role.scope == diesel_models::enums::RoleScope::Organization)
+            .find(|role| {
+                let org_id = entity_type.get_organization_id();
+                let merchant_id = entity_type.get_merchant_id();
+                let profile_id = entity_type.get_profile_id();
+
+                let condition = org_id
+                    .as_ref()
+                    .map(|org_id| {
+                        *org_id == role.org_id
+                            && role.entity_type == EntityType::Organization
+                            && role.scope == RoleScope::Organization
+                    })
+                    .unwrap_or(true)
+                    || merchant_id
+                        .as_ref()
+                        .map(|merchant_id| {
+                            let entity_in_vec = [EntityType::Merchant, EntityType::Organization];
+                            *merchant_id == role.merchant_id
+                                && role.scope == RoleScope::Merchant
+                                && entity_in_vec.contains(&role.entity_type)
+                        })
+                        .unwrap_or(true)
+                    || profile_id
+                        .as_ref()
+                        .zip(role.profile_id.as_ref())
+                        .map(|(profile_id, role_profile_id)| {
+                            let entity_in_vec = [
+                                EntityType::Profile,
+                                EntityType::Merchant,
+                                EntityType::Organization,
+                            ];
+                            profile_id == role_profile_id
+                                && role.scope == RoleScope::Profile
+                                && entity_in_vec.contains(&role.entity_type)
+                        })
+                        .unwrap_or(true);
+
+                role.role_name == role_name && condition
             })
-            .cloned()
-            .collect();
-
-        if roles_list.is_empty() {
-            return Err(errors::StorageError::ValueNotFound(format!(
-                "No role found for merchant id = {:?} and org_id = {:?}",
-                merchant_id, org_id
-            ))
-            .into());
-        }
+            .cloned();
 
-        Ok(roles_list)
+        roles_list.ok_or(
+            errors::StorageError::ValueNotFound(
+                "No role available with role_name={role_name} in org_id = {org_id:?}".to_string(),
```

Comment:
Maybe printing the `EntityTypePayload` is more appropriate here.

Reply:
Removed the query as not needed anymore